File: runtime/ffi/go/avltree/avltree.go
package avltree

import (
	"encoding/json"
	"strings"
)

// Node is a node in an AVL tree storing ints.
type Node struct {
	Data    int      `json:"Data"`
	Balance int      `json:"Balance"`
	Link    [2]*Node `json:"Link"`
}

func opp(dir int) int { return 1 - dir }

func single(root *Node, dir int) *Node {
	save := root.Link[opp(dir)]
	root.Link[opp(dir)] = save.Link[dir]
	save.Link[dir] = root
	return save
}

func double(root *Node, dir int) *Node {
	save := root.Link[opp(dir)].Link[dir]

	root.Link[opp(dir)].Link[dir] = save.Link[opp(dir)]
	save.Link[opp(dir)] = root.Link[opp(dir)]
	root.Link[opp(dir)] = save

	save = root.Link[opp(dir)]
	root.Link[opp(dir)] = save.Link[dir]
	save.Link[dir] = root
	return save
}

func adjustBalance(root *Node, dir, bal int) {
	n := root.Link[dir]
	nn := n.Link[opp(dir)]
	switch nn.Balance {
	case 0:
		root.Balance = 0
		n.Balance = 0
	case bal:
		root.Balance = -bal
		n.Balance = 0
	default:
		root.Balance = 0
		n.Balance = bal
	}
	nn.Balance = 0
}

func insertBalance(root *Node, dir int) *Node {
	n := root.Link[dir]
	bal := 2*dir - 1
	if n.Balance == bal {
		root.Balance = 0
		n.Balance = 0
		return single(root, opp(dir))
	}
	adjustBalance(root, dir, bal)
	return double(root, opp(dir))
}

func insertR(root *Node, data int) (*Node, bool) {
	if root == nil {
		return &Node{Data: data}, false
	}
	dir := 0
	if root.Data < data {
		dir = 1
	}
	var done bool
	root.Link[dir], done = insertR(root.Link[dir], data)
	if done {
		return root, true
	}
	root.Balance += 2*dir - 1
	switch root.Balance {
	case 0:
		return root, true
	case 1, -1:
		return root, false
	}
	return insertBalance(root, dir), true
}

// Insert inserts data into the tree.
func Insert(tree **Node, data int) { *tree, _ = insertR(*tree, data) }

func removeBalance(root *Node, dir int) (*Node, bool) {
	n := root.Link[opp(dir)]
	bal := 2*dir - 1
	switch n.Balance {
	case -bal:
		root.Balance = 0
		n.Balance = 0
		return single(root, dir), false
	case bal:
		adjustBalance(root, opp(dir), -bal)
		return double(root, dir), false
	}
	root.Balance = -bal
	n.Balance = bal
	return single(root, dir), true
}

func removeR(root *Node, data int) (*Node, bool) {
	if root == nil {
		return nil, false
	}
	if root.Data == data {
		switch {
		case root.Link[0] == nil:
			return root.Link[1], false
		case root.Link[1] == nil:
			return root.Link[0], false
		}
		heir := root.Link[0]
		for heir.Link[1] != nil {
			heir = heir.Link[1]
		}
		root.Data = heir.Data
		data = heir.Data
	}
	dir := 0
	if root.Data < data {
		dir = 1
	}
	var done bool
	root.Link[dir], done = removeR(root.Link[dir], data)
	if done {
		return root, true
	}
	root.Balance += 1 - 2*dir
	switch root.Balance {
	case 1, -1:
		return root, true
	case 0:
		return root, false
	}
	return removeBalance(root, dir)
}

// Remove removes data from the tree.
func Remove(tree **Node, data int) { *tree, _ = removeR(*tree, data) }

// Dump returns a JSON representation with indentation.
func Dump(tree *Node) string {
	b, _ := json.MarshalIndent(tree, "", "   ")
	return string(b)
}

// ExampleOutput returns the output for the Rosetta task.
func ExampleOutput() string {
	var tree *Node
	var sb strings.Builder
	sb.WriteString("Empty tree:\n")
	sb.WriteString(Dump(tree))
	sb.WriteString("\n\nInsert test:\n")
	Insert(&tree, 3)
	Insert(&tree, 1)
	Insert(&tree, 4)
	Insert(&tree, 1)
	Insert(&tree, 5)
	sb.WriteString(Dump(tree))
	sb.WriteString("\n\nRemove test:\n")
	Remove(&tree, 3)
	Remove(&tree, 1)
	sb.WriteString(Dump(tree))
	return sb.String()
}

File: tests/rosetta/transpiler/php/avl-tree.php
<?php
ini_set('memory_limit', '-1');
function Node($data) {
  global $getLink, $setLink, $opp, $single, $double, $adjustBalance, $insertBalance, $insertR, $Insert, $removeBalance, $removeR, $Remove, $indentStr, $dumpNode, $dump, $main;
  return ['Data' => $data, 'Balance' => 0, 'Link' => [null, null]];
}
function getLink($n, $dir) {
  global $Node, $setLink, $opp, $single, $double, $adjustBalance, $insertBalance, $insertR, $Insert, $removeBalance, $removeR, $Remove, $indentStr, $dumpNode, $dump, $main;
  return ($n['Link'])[$dir];
}
function setLink($n, $dir, $v) {
  global $Node, $getLink, $opp, $single, $double, $adjustBalance, $insertBalance, $insertR, $Insert, $removeBalance, $removeR, $Remove, $indentStr, $dumpNode, $dump, $main;
  $links = $n['Link'];
  $links[$dir] = $v;
  $n['Link'] = $links;
}
function opp($dir) {
  global $Node, $getLink, $setLink, $single, $double, $adjustBalance, $insertBalance, $insertR, $Insert, $removeBalance, $removeR, $Remove, $indentStr, $dumpNode, $dump, $main;
  return 1 - $dir;
}
function single($root, $dir) {
  global $Node, $getLink, $setLink, $opp, $double, $adjustBalance, $insertBalance, $insertR, $Insert, $removeBalance, $removeR, $Remove, $indentStr, $dumpNode, $dump, $main;
  $tmp = getLink($root, opp($dir));
  setLink($root, opp($dir), getLink($tmp, $dir));
  setLink($tmp, $dir, $root);
  return $tmp;
}
function double($root, $dir) {
  global $Node, $getLink, $setLink, $opp, $single, $adjustBalance, $insertBalance, $insertR, $Insert, $removeBalance, $removeR, $Remove, $indentStr, $dumpNode, $dump, $main;
  $tmp = getLink(getLink($root, opp($dir)), $dir);
  setLink(getLink($root, opp($dir)), $dir, getLink($tmp, opp($dir)));
  setLink($tmp, opp($dir), getLink($root, opp($dir)));
  setLink($root, opp($dir), $tmp);
  $tmp = getLink($root, opp($dir));
  setLink($root, opp($dir), getLink($tmp, $dir));
  setLink($tmp, $dir, $root);
  return $tmp;
}
function adjustBalance($root, $dir, $bal) {
  global $Node, $getLink, $setLink, $opp, $single, $double, $insertBalance, $insertR, $Insert, $removeBalance, $removeR, $Remove, $indentStr, $dumpNode, $dump, $main;
  $n = getLink($root, $dir);
  $nn = getLink($n, opp($dir));
  if ($nn['Balance'] == 0) {
  $root['Balance'] = 0;
  $n['Balance'] = 0;
} else {
  if ($nn['Balance'] == $bal) {
  $root['Balance'] = -$bal;
  $n['Balance'] = 0;
} else {
  $root['Balance'] = 0;
  $n['Balance'] = $bal;
};
}
  $nn['Balance'] = 0;
}
function insertBalance($root, $dir) {
  global $Node, $getLink, $setLink, $opp, $single, $double, $adjustBalance, $insertR, $Insert, $removeBalance, $removeR, $Remove, $indentStr, $dumpNode, $dump, $main;
  $n = getLink($root, $dir);
  $bal = 2 * $dir - 1;
  if ($n['Balance'] == $bal) {
  $root['Balance'] = 0;
  $n['Balance'] = 0;
  return single($root, opp($dir));
}
  adjustBalance($root, $dir, $bal);
  return double($root, opp($dir));
}
function insertR($root, $data) {
  global $Node, $getLink, $setLink, $opp, $single, $double, $adjustBalance, $insertBalance, $Insert, $removeBalance, $removeR, $Remove, $indentStr, $dumpNode, $dump, $main;
  if ($root == null) {
  return ['node' => Node($data), 'done' => false];
}
  $node = $root;
  $dir = 0;
  if ((ord($node['Data'])) < $data) {
  $dir = 1;
}
  $r = insertR(getLink($node, $dir), $data);
  setLink($node, $dir, $r['node']);
  if ($r['done']) {
  return ['node' => $node, 'done' => true];
}
  $node['Balance'] = (ord($node['Balance'])) + (2 * $dir - 1);
  if ($node['Balance'] == 0) {
  return ['node' => $node, 'done' => true];
}
  if ($node['Balance'] == 1 || $node['Balance'] == (-1)) {
  return ['node' => $node, 'done' => false];
}
  return ['node' => insertBalance($node, $dir), 'done' => true];
}
function Insert($tree, $data) {
  global $Node, $getLink, $setLink, $opp, $single, $double, $adjustBalance, $insertBalance, $insertR, $removeBalance, $removeR, $Remove, $indentStr, $dumpNode, $dump, $main;
  $r = insertR($tree, $data);
  return $r['node'];
}
function removeBalance($root, $dir) {
  global $Node, $getLink, $setLink, $opp, $single, $double, $adjustBalance, $insertBalance, $insertR, $Insert, $removeR, $Remove, $indentStr, $dumpNode, $dump, $main;
  $n = getLink($root, opp($dir));
  $bal = 2 * $dir - 1;
  if ($n['Balance'] == (-$bal)) {
  $root['Balance'] = 0;
  $n['Balance'] = 0;
  return ['node' => single($root, $dir), 'done' => false];
}
  if ($n['Balance'] == $bal) {
  adjustBalance($root, opp($dir), (-$bal));
  return ['node' => double($root, $dir), 'done' => false];
}
  $root['Balance'] = -$bal;
  $n['Balance'] = $bal;
  return ['node' => single($root, $dir), 'done' => true];
}
function removeR($root, $data) {
  global $Node, $getLink, $setLink, $opp, $single, $double, $adjustBalance, $insertBalance, $insertR, $Insert, $removeBalance, $Remove, $indentStr, $dumpNode, $dump, $main;
  if ($root == null) {
  return ['node' => null, 'done' => false];
}
  $node = $root;
  if ((ord($node['Data'])) == $data) {
  if (getLink($node, 0) == null) {
  return ['node' => getLink($node, 1), 'done' => false];
};
  if (getLink($node, 1) == null) {
  return ['node' => getLink($node, 0), 'done' => false];
};
  $heir = getLink($node, 0);
  while (getLink($heir, 1) != null) {
  $heir = getLink($heir, 1);
};
  $node['Data'] = $heir['Data'];
  $data = ord($heir['Data']);
}
  $dir = 0;
  if ((ord($node['Data'])) < $data) {
  $dir = 1;
}
  $r = removeR(getLink($node, $dir), $data);
  setLink($node, $dir, $r['node']);
  if ($r['done']) {
  return ['node' => $node, 'done' => true];
}
  $node['Balance'] = (ord($node['Balance'])) + 1 - 2 * $dir;
  if ($node['Balance'] == 1 || $node['Balance'] == (-1)) {
  return ['node' => $node, 'done' => true];
}
  if ($node['Balance'] == 0) {
  return ['node' => $node, 'done' => false];
}
  return removeBalance($node, $dir);
}
function Remove($tree, $data) {
  global $Node, $getLink, $setLink, $opp, $single, $double, $adjustBalance, $insertBalance, $insertR, $Insert, $removeBalance, $removeR, $indentStr, $dumpNode, $dump, $main;
  $r = removeR($tree, $data);
  return $r['node'];
}
function indentStr($n) {
  global $Node, $getLink, $setLink, $opp, $single, $double, $adjustBalance, $insertBalance, $insertR, $Insert, $removeBalance, $removeR, $Remove, $dumpNode, $dump, $main;
  $s = '';
  $i = 0;
  while ($i < $n) {
  $s = $s . ' ';
  $i = $i + 1;
};
  return $s;
}
function dumpNode($node, $indent, $comma) {
  global $Node, $getLink, $setLink, $opp, $single, $double, $adjustBalance, $insertBalance, $insertR, $Insert, $removeBalance, $removeR, $Remove, $indentStr, $dump, $main;
  $sp = indentStr($indent);
  if ($node == null) {
  $line = $sp . 'null';
  if ($comma) {
  $line = $line . ',';
};
  echo rtrim($line), PHP_EOL;
} else {
  echo rtrim($sp . '{'), PHP_EOL;
  echo rtrim(indentStr($indent + 3) . '"Data": ' . json_encode($node['Data'], 1344) . ','), PHP_EOL;
  echo rtrim(indentStr($indent + 3) . '"Balance": ' . json_encode($node['Balance'], 1344) . ','), PHP_EOL;
  echo rtrim(indentStr($indent + 3) . '"Link": ['), PHP_EOL;
  dumpNode(getLink($node, 0), $indent + 6, true);
  dumpNode(getLink($node, 1), $indent + 6, false);
  echo rtrim(indentStr($indent + 3) . ']'), PHP_EOL;
  $end = $sp . '}';
  if ($comma) {
  $end = $end . ',';
};
  echo rtrim($end), PHP_EOL;
}
}
function dump($node, $indent) {
  global $Node, $getLink, $setLink, $opp, $single, $double, $adjustBalance, $insertBalance, $insertR, $Insert, $removeBalance, $removeR, $Remove, $indentStr, $dumpNode, $main;
  dumpNode($node, $indent, false);
}
function main() {
  global $Node, $getLink, $setLink, $opp, $single, $double, $adjustBalance, $insertBalance, $insertR, $Insert, $removeBalance, $removeR, $Remove, $indentStr, $dumpNode, $dump;
  $tree = null;
  echo rtrim('Empty tree:'), PHP_EOL;
  dump($tree, 0);
  echo rtrim(''), PHP_EOL;
  echo rtrim('Insert test:'), PHP_EOL;
  $tree = Insert($tree, 3);
  $tree = Insert($tree, 1);
  $tree = Insert($tree, 4);
  $tree = Insert($tree, 1);
  $tree = Insert($tree, 5);
  dump($tree, 0);
  echo rtrim(''), PHP_EOL;
  echo rtrim('Remove test:'), PHP_EOL;
  $tree = Remove($tree, 3);
  $tree = Remove($tree, 1);
  $t = $tree;
  $t['Balance'] = 0;
  $tree = $t;
  dump($tree, 0);
}
main();
